marklogic-client-api/src/main/java/com/marklogic/client/impl/SchemaDefImpl.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2022 MarkLogic Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.marklogic.client.impl;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.marklogic.client.expression.SchemaDefExpr;
+
+public class SchemaDefImpl implements SchemaDefExpr, BaseTypeImpl.BaseArgImpl {
+    private String kind;
+    private String mode;
+    private String schemaUri;
+
+    public SchemaDefImpl(String kind){
```

Comment:
I think schemaUri should be required, and "kind" seems optional. And this class can make a very educated guess to determine what "kind" should be based on schemaUri:

1. If schemaUri ends with .mjs, then kind = mjs
2. If schemaUri ends with .xsl or .xslt, then kind = xslt

The javadocs for the `schemaDef(String schemaUri)` method can make this clear - i.e. by telling the user that they don't need to call `withKind` unless their schemaUri doesn't end with mjs, xsl, or xslt. That provides the best UX for the user, such that they only need to call `schemaDef(String schemaUri)` and then they can optionally call `withMode` as well.

Reply:
Oops - I was thinking transformDoc! But the same logic applies, with checking for common extensions of json/xsd/sch.

Reply:
Will revisit this after the design is finalized.